common/txmgr/txmgr.go

```diff
@@ -625,6 +640,52 @@ func (b *Txm[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) CountTrans
 	return b.txStore.CountTransactionsByState(ctx, state, b.chainID)
 }
 
+func (b *Txm[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) GetTransactionStatus(ctx context.Context, transactionID uuid.UUID) (status commontypes.TransactionStatus, err error) {
+	// Loads attempts and receipts in the transaction
+	tx, err := b.txStore.FindTxWithIdempotencyKey(ctx, transactionID.String(), b.chainID)
+	if err != nil {
+		return status, fmt.Errorf("failed to find transaction with IdempotencyKey %s: %w", transactionID.String(), err)
+	}
+	// This check is required since a no-rows error returns nil err
+	if tx == nil {
+		return status, fmt.Errorf("failed to find transaction with IdempotencyKey %s", transactionID.String())
+	}
+	switch tx.State {
+	case TxUnconfirmed, TxConfirmedMissingReceipt:
+		// Return unconfirmed for ConfirmedMissingReceipt since a receipt is required to determine if it is finalized
+		return commontypes.Unconfirmed, nil
+	case TxConfirmed:
+		var receipt txmgrtypes.ChainReceipt[TX_HASH, BLOCK_HASH]
+		// Find tx receipt if one exists
+		for _, attempt := range tx.TxAttempts {
+			if len(attempt.Receipts) > 0 {
+				// Tx will only have one receipt
+				receipt = attempt.Receipts[0]
+				break
+			}
+		}
+		b.finalizedBlockNumMu.RLock()
+		defer b.finalizedBlockNumMu.RUnlock()
+		if receipt != nil && b.latestFinalizedBlockNum != 0 && receipt.GetBlockNumber().Cmp(big.NewInt(b.latestFinalizedBlockNum)) <= 0 {
+			// Return finalized if tx receipt's block is equal or older than the latest finalized block
+			return commontypes.Finalized, nil
+		}
+		// Return unconfirmed if tx receipt's block is newer than the latest finalized block
+		return commontypes.Unconfirmed, nil
+	case TxFatalError:
+		// Use the TxError builder to classify the error message stored for the tx
+		txErr := b.newTxError(tx.GetError())
+		if txErr != nil && txErr.IsFatal() {
+			return commontypes.Fatal, tx.GetError()
+		}
+		// Return failed for all other tx's marked as FatalError
+		return commontypes.Failed, tx.GetError()
```

Comment:
This `newTxError` indirection is really difficult to read. Can we try to use the error APIs directly here?
```suggestion
		// Use the TxError builder to classify the error message stored for the tx
		status := commontypes.Failed
		err := tx.GetError()
		var txErr ErrType
		if ok := errors.As(err, &txErr); ok && txErr.IsFatal() {
			status = commontypes.Fatal
		}
		return status, err
```

Reply:
I don't think the `errors.As` will successfully match the type since `tx.GetError()` returns the basic Go error type while the `newTxError` is building a custom error type `txmgrtypes.TxError`. Maybe the better fix is to set the `Tx.Error` field to a wrapped error type to use directly but think that change would propagate out of the scope of this ticket.

Reply:
Ah, I misunderstood what is happening here. How about just renaming the field like the comment, e.g. `b.classifyError(err)`

Reply:
I'm ok with the rename if it makes this clearer but the `newTxError` method is actually just building the error type that allows us to classify errors. The classification is actually happening at `txErr.IsFatal()`. The comment could be worded better to explain that. Maybe `TxError` also isn't descriptive enough. I can rename this to `ErrorClassifier` and the `newTxError` method to `newErrorClassifier`.